File: frontend/src/types/memory.ts
/** Browser-facing contracts for reviewable memories and explicit scopes. */

export type PlatformFeatureFlags = {
  memory_enabled: boolean
  rag_enabled: boolean
}

export type MemoryScope = 'private' | `household:${number}` | `lineage:${number}`
export type MemoryScopeKind = 'private' | 'household' | 'lineage'
export type MemorySensitivity = 'normal' | 'sensitive' | 'high' | 'local_required'
export type MemoryCandidateStatus = 'pending' | 'dismissed' | 'confirmed'
export type MemoryStatus = 'active' | 'revoked' | 'deleted'
export type MemoryScopeSelection = MemoryScope
export type MemorySourceStatus = 'available' | 'deleted_snapshot' | 'unavailable' | 'unverified'
export type MemorySourceKind = 'manual' | 'agent_message' | 'rag_chunk' | 'legacy'

export type MemoryCandidateSource =
  | { kind: 'manual' }
  | { kind: 'agent_message'; message_id: number }
  | {
      kind: 'rag_chunk'
      document_id: number
      chunk_id: number
      revision: number
      index_version: string
      /** The space in which the user searched, not the source document's scope. */
      space_id: number
    }

export const MEMORY_SOURCE_STATUS_LABELS: Record<MemorySourceStatus, string> = {
  available: '来源可用',
  deleted_snapshot: '原会话已删除，保留已确认快照',
  unavailable: '来源不可用，内容暂不显示',
  unverified: '来源待验证，内容暂不显示',
}

export function memorySourceReadable(status: MemorySourceStatus): boolean {
  return status === 'available' || status === 'deleted_snapshot'
}

export const MEMORY_CANDIDATE_STATUS_LABELS: Record<MemoryCandidateStatus, string> = {
  pending: '待确认',
  dismissed: '已忽略',
  confirmed: '已确认',
}

export const MEMORY_SCOPE_LABELS: Record<MemoryScopeKind | 'private', string> = {
  private: '仅我可见',
  household: '家庭共享',
  lineage: '族谱共享',
}

export const MEMORY_SENSITIVITY_LABELS: Record<MemorySensitivity, string> = {
  normal: '普通',
  sensitive: '敏感',
  high: '高敏感',
  local_required: '仅本地模型',
}


export interface MemoryCitation {
  source_type: string
  source_id: string
  scope: string
  sensitivity: string
  revision: number
  citation_handle: string
  text?: string
  chunk_id?: number
  document_id?: number
  index_version?: string
}

export interface RagSearchResult extends MemoryCitation {
  chunk_id: number
  document_id: number
  text: string
  index_version: string
  sensitivity: MemorySensitivity
  space_id: number | null
  allowed_scopes: MemoryScope[]
}

/** Shared scopes always carry the selected space id at the request boundary. */
export interface MemoryCandidate {
  id: number
  source_message_id: number | null
  source_document_ref: string | null
  source_span_json: Record<string, unknown>
  source_kind: MemorySourceKind
  source_status: MemorySourceStatus
  allowed_scopes: MemoryScope[]
  raw_quote: string | null
  summary: string | null
  suggested_scope: MemoryScopeKind
  purpose: string | null
  sensitivity: MemorySensitivity
  extractor_version: string
  status: MemoryCandidateStatus
  memory_id: number | null
  created_at: string
  decided_at: string | null
}

export interface Memory {
  id: number
  source_candidate_id: number | null
  source_message_id: number | null
  source_document_ref: string | null
  source_kind: MemorySourceKind
  source_status: MemorySourceStatus
  allowed_scopes: MemoryScope[]
  raw_quote: string | null
  content: string | null
  purpose: string | null
  scope: MemoryScopeKind
  space_id: number | null
  sensitivity: MemorySensitivity
  confirmation_status: 'confirmed'
  revision: number
  retention_until: string | null
  status: MemoryStatus
  revoked_at: string | null
  created_at: string
  updated_at: string
}

File: backend/app/schemas/memory.py
"""V2.5 Memory and RAG HTTP contracts."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class ManualMemorySource(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: Literal["manual"]


class AgentMessageMemorySource(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: Literal["agent_message"]
    message_id: int = Field(gt=0)


class RAGChunkMemorySource(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: Literal["rag_chunk"]
    document_id: int = Field(gt=0)
    chunk_id: int = Field(gt=0)
    revision: int = Field(gt=0)
    index_version: str = Field(min_length=1, max_length=32, pattern=r"^[A-Za-z0-9_.-]+$")
    space_id: int = Field(gt=0)


MemorySource = Annotated[
    ManualMemorySource | AgentMessageMemorySource | RAGChunkMemorySource,
    Field(discriminator="kind"),
]


class MemoryCandidateCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    source: MemorySource | None = None
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=128)
    source_message_id: int | None = Field(default=None, gt=0)
    source_document_ref: str | None = Field(default=None, max_length=255)
    source_span: dict[str, Any] = Field(default_factory=dict)
    raw_quote: str | None = Field(default=None, min_length=1, max_length=12_000)
    summary: str = Field(min_length=1, max_length=20_000)
    suggested_scope: Literal["private", "household", "lineage"] = "private"
    purpose: str = Field(min_length=1, max_length=120)
    sensitivity: Literal["normal", "sensitive", "high", "local_required"] = "normal"


class MemoryCandidateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_message_id: int | None
    source_document_ref: str | None
    source_span_json: dict[str, Any]
    source_kind: Literal["manual", "agent_message", "rag_chunk", "legacy"]
    source_status: Literal["available", "deleted_snapshot", "unavailable", "unverified"]
    allowed_scopes: list[str]
    raw_quote: str | None
    summary: str | None
    suggested_scope: str
    purpose: str | None
    sensitivity: str
    extractor_version: str
    status: Literal["pending", "dismissed", "confirmed"]
    memory_id: int | None
    created_at: datetime
    decided_at: datetime | None


class MemoryConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    scope: str = Field(min_length=1, max_length=64)
    retention_days: int | None = Field(default=None, ge=1, le=3650)


class MemoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_candidate_id: int | None
    source_message_id: int | None
    source_document_ref: str | None
    source_span_json: dict[str, Any]
    source_kind: Literal["manual", "agent_message", "rag_chunk", "legacy"]
    source_status: Literal["available", "deleted_snapshot", "unavailable", "unverified"]
    allowed_scopes: list[str]
    raw_quote: str | None
    content: str | None
    scope: str
    space_id: int | None
    sensitivity: str
    purpose: str | None
    confirmation_status: Literal["confirmed"]
    revision: int
    retention_until: datetime | None
    status: str
    revoked_at: datetime | None
    created_at: datetime
    updated_at: datetime


class RAGSearchOut(BaseModel):
    chunk_id: int
    document_id: int
    source_type: str
    source_id: str
    text: str
    scope: str
    sensitivity: str
    revision: int
    index_version: str
    citation_handle: str
    space_id: int | None
    allowed_scopes: list[str]
